### PR TITLE
Kate / DTRA-378 / TS migration: Multiplier components and containers in Trader package

### PR DESCRIPTION
**Pull request by kate-deriv**
## Changes:

- TS migration
- Refactoring

### Screenshots:

Please provide some screenshots of the change.

### CODE REVIEW

File: packages/components/src/components/dialog/dialog.tsx [hunk 1]
```diff
@@ -98,7 +98,7 @@ const Dialog = ({
         }
     };
 
-    const validateClickOutside = () => dismissable || (has_close_icon && is_visible && is_closed_on_cancel);
+    const validateClickOutside = () => !!dismissable || !!(has_close_icon && is_visible && is_closed_on_cancel);
```
Comment (kate-deriv) [inline packages/components/src/components/dialog/dialog.tsx:101]: Function `validateClickOutside` is passed as an argument inside of `useOnClickOutside` hook. This hook is expecting function, which is returning `boolean`.

File: packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-amount-modal.tsx [hunk 2]
```diff
@@ -5,26 +5,28 @@ import { localize, Localize } from '@deriv/translations';
...
-    const {
-        ui: { enableApp, disableApp },
-    } = useStore();
 
     return (
         <React.Fragment>
             <Modal
                 id='dt_trade_parameters_mobile'
                 className='trade-params'
-                enableApp={enableApp}
```
Comment (kate-deriv) [inline packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-amount-modal.tsx:23]: `<Modal />` is not expecting `enableApp` and `disableApp` props

File: packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-amount-modal.tsx [hunk 3]
```diff
@@ -76,7 +78,7 @@ const TradeParamsMobile = observer(({ toggleModal }) => {
         // eslint-disable-next-line react-hooks/exhaustive-deps
     }, [stake_value]);
 
-    const setSelectedAmount = (basis, stake) => {
```
Comment (kate-deriv) [inline packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-amount-modal.tsx:79]: Parameter `basis` wasn't used inside of the function, so I deleted it here and in places where that function was called.

File: packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-amount-modal.tsx [hunk 4]
```diff
@@ -104,10 +106,10 @@ const TradeParamsMobile = observer(({ toggleModal }) => {
                 toggleModal={toggleModal}
                 amount_tab_idx={0}
                 setSelectedAmount={setSelectedAmount}
-                setAmountError={() => {}}
```
Comment (kate-deriv) [inline packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-amount-modal.tsx:107]: Made this param optional in order not to pass 'empty' function

File: packages/trader/src/Modules/Trading/Containers/Multiplier/risk-management-dialog.tsx [hunk 5]
```diff
@@ -83,9 +93,9 @@ const RiskManagementDialog = observer(({ is_open, onClose, toggleDialog }) => {
         toggleDialog();
     };
 
-    const resetAndClose = (...args) => {
+    const resetAndClose = () => {
         setState(applied_risk_management_state);
-        onClose(...args);
+        onClose();
```
Comment (kate-deriv) [inline packages/trader/src/Modules/Trading/Containers/Multiplier/risk-management-dialog.tsx:98]: `onClose` function is actually a `() => setDialogOpen(false)` (passed as a props from _risk-management-info.tsx_), so there is no need in `...args`

File: packages/trader/src/Modules/Trading/Components/Form/TradeParams/amount-mobile.tsx [hunk 6]
```diff
@@ -127,11 +127,14 @@ const Basis = observer(
     }
 );
 
-type TAmountMobile = React.ComponentProps<typeof Basis> & {
```
Comment (kate-deriv) [inline packages/trader/src/Modules/Trading/Components/Form/TradeParams/amount-mobile.tsx:130]: We don't need all props from `<Basis />`

File: packages/trader/src/Modules/Trading/Containers/trade-params-mobile.tsx [hunk 7]
```diff
@@ -314,8 +314,6 @@ const TradeParamsMobile = observer(
                             setAmountError={setAmountError}
                             stake_value={stake_value}
                             payout_value={payout_value}
-                            basis={''}
```
Comment (kate-deriv) [inline packages/trader/src/Modules/Trading/Containers/trade-params-mobile.tsx:317]: Made this params optional

File: packages/trader/src/Modules/Trading/Containers/Multiplier/risk-management-dialog.tsx [hunk 8]
```diff
@@ -63,15 +73,15 @@ const RiskManagementDialog = observer(({ is_open, onClose, toggleDialog }) => {
         });
     };
 
-    const onChange = e => {
+    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
         const { name, value } = e.target;
         const new_state = { ...state };
-        new_state[name] = value;
+        new_state[name as 'take_profit' | 'stop_loss' | 'cancellation_duration'] = value;
```
Comment (maryia-deriv) [inline packages/trader/src/Modules/Trading/Containers/Multiplier/risk-management-dialog.tsx:79]:
```suggestion
        new_state[name as keyof typeof new_state] = value;
```
will this work?)

File: packages/trader/src/Types/common-prop.type.ts [hunk 9]
```diff
@@ -30,4 +30,22 @@ export type TError = {
     };
 };
 
+export type TOnProposalResponse = {
```
Comment (maryia-deriv) [inline packages/trader/src/Types/common-prop.type.ts:33]: @kate-deriv let's use existing type instead: `TTradeStore['onProposalResponse']`

File: packages/trader/src/Stores/Modules/Trading/trade-store.ts [hunk 10]
```diff
@@ -212,6 +212,7 @@ export default class TradeStore extends BaseStore {
     basis_list: Array<TTextValueStrings> = [];
     currency = '';
     stake_boundary: Partial<TStakeBoundary> = {};
+    trade_stop_out = '';
```
Comment (maryia-deriv) [inline packages/trader/src/Stores/Modules/Trading/trade-store.ts:215]: @kate-deriv please remove it, it's unused, and please also remove it from packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-amount-modal.jsx

File: packages/trader/src/Modules/Trading/Components/Form/TradeParams/amount-mobile.tsx [hunk 11]
```diff
@@ -9,13 +9,13 @@ import MinMaxStakeInfo from './min-max-stake-info';
 
 type TBasis = {
     basis: string;
```
Comment (maryia-deriv) [inline packages/trader/src/Modules/Trading/Components/Form/TradeParams/amount-mobile.tsx:11]: @kate-deriv is it supposed to be optional? you mentioned it here: https://github.com/kate-deriv/deriv-app/pull/26/files#r1311302301

File: packages/trader/src/Modules/Trading/Components/Elements/Multiplier/cancel-deal-mobile.tsx [hunk 12]
```diff
@@ -78,7 +101,7 @@ const CancelDeal = observer(
                             id='dt_multiplier-stake__tooltip'
                             is_bubble_hover_enabled
                             classNameBubble='trade-container__deal-cancellation-popover'
-                            zIndex={9999}
+                            zIndex={'9999'}
```
Comment (maryia-deriv) [inline packages/trader/src/Modules/Trading/Components/Elements/Multiplier/cancel-deal-mobile.tsx:104]:
```suggestion
                            zIndex='9999'
```

File: packages/trader/src/Modules/Trading/Components/Elements/Multiplier/cancel-deal-mobile.tsx [hunk 13]
```diff
@@ -6,38 +6,61 @@ import Fieldset from 'App/Components/Form/fieldset.jsx';
...
+                enableApp={enableApp}
+                portal_element_id='modal_root'
+            >
+                <Text size='xxxs' color='general'>
+                    <Localize i18n_default_text='Take profit and/or stop loss are not available while deal cancellation is active.' />
+                </Text>
+                <Checkbox
+                    defaultChecked={!should_show_cancellation_warning}
+                    onChange={() => toggleCancellationWarning()}
+                    name={'should_show_cancellation_warning'}
```
Comment (maryia-deriv) [inline packages/trader/src/Modules/Trading/Components/Elements/Multiplier/cancel-deal-mobile.tsx:54]:
```suggestion
                    name='should_show_cancellation_warning'
```
nit-pick :)

File: packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-options.tsx [hunk 14]
```diff
@@ -6,21 +6,26 @@ import { localize } from '@deriv/translations';
...
-    const [stop_out, setStopOut] = React.useState(null);
+    const [commission, setCommission] = React.useState<string | number | null>(null);
+    const [stop_out, setStopOut] = React.useState<string | number | null>(null);
     const isMounted = useIsMounted();
 
     React.useEffect(() => {
         if (!amount) return undefined;
 
-        const onProposalResponse = ({ echo_req, proposal, subscription }) => {
+        const onProposalResponse = ({ echo_req, proposal, subscription }: TOnProposalResponse) => {
```
Comment (maryia-deriv) [inline packages/trader/src/Modules/Trading/Containers/Multiplier/multiplier-options.tsx:25]:
```suggestion
        const onProposalResponse: TTradeStore['onProposalResponse'] = ({ echo_req, proposal, subscription }) => {
```

File: packages/trader/src/Modules/Trading/Containers/Multiplier/risk-management-dialog.tsx (hunk 8 above)
Comment (kate-deriv) [reply packages/trader/src/Modules/Trading/Containers/Multiplier/risk-management-dialog.tsx:79]: No, it wouldn't :(
Reason: `new_state` was created based on `state` object, which type is ⤵️
```
state: {
    take_profit: string | undefined;
    stop_loss: string | undefined;
    has_take_profit: boolean;
    has_stop_loss: boolean;
    has_cancellation: boolean;
    cancellation_duration: string;
}
```
So, some of the fields are `boolean` and a `value` (from `e.target`) is a `string`. And here comes an error, that `string` is not assignable to that type (which includes `boolean`). That is why I used union type

**Review by maryia-deriv — APPROVED** [review]

**Review by henry-deriv — APPROVED** [review]

**Pull request merged**